src/wallet/wallet2.cpp

```diff
@@ -3604,6 +3719,19 @@ void wallet2::load(const std::string& wallet_, const epee::wipeable_string& pass
     add_subaddress_account(tr("Primary account"));
 
   m_local_bc_height = m_blockchain.size();
+
+  if (!m_ring_history_saved)
+  {
+    try
+    {
+      find_and_save_rings();
+      m_ring_history_saved = true;
```

Comment:
IIRC I had a reason for this, but I can't recall it right now. I'll leave it that way for now and go back to it if I remember.

Reply:
But this makes simplewallet not update the blackball db even when the user typed `save_known_rings`. Is that fine?

Reply:
Or, is `m_ring_history_saved` needed in the first place? The user can explicitly invoke the db generation at any time by typing `save_known_rings`.

Reply:
It's useful to avoid querying the daemon every time with the set of txes the user sent, which if a third party daemon is used at some point, would leak info. Also avoids all the unneeded work, which I suspect can get quite significant for large wallets (ie, exchanges). Since we can do it automatically without the user having to know they need to run save_known_rings once, it's better to do so.

Reply:
I see.

Reply:
After some more thought, I'll do what you suggested originally and move the flag set to the function on success.